src/stream/from_fn.rs

```diff
@@ -0,0 +1,104 @@
+use futures::stream::Stream;
+use pin_utils::{unsafe_pinned, unsafe_unpinned};
+
+use std::fmt;
+use std::future::Future;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+/// Creates a new stream where each iteration calls the provided closure.
+///
+/// This allows creating a custom stream with any behavior
+/// without using the more verbose syntax of creating a dedicated type
+/// and implementing the `Stream` trait for it.
+///
+/// Note that the `FromFn` stream doesn’t make assumptions about the behavior of the closure,
+/// and therefore conservatively does not implement [`FusedStream`](futures_core::stream::FusedStream).
+///
+/// The closure can use captures and its environment to track state across iterations. Depending on
+/// how the stream is used, this may require specifying the `move` keyword on the closure.
+///
+/// # Examples
+///
+/// ```
+/// # futures::executor::block_on(async {
+/// use async_std::{future, stream};
+/// use std::sync::atomic::{AtomicUsize, Ordering};
+///
+/// static COUNT: AtomicUsize = AtomicUsize::new(0);
+/// let stream = stream::from_fn(|| {
+///     // Increment our count. This is why we started at zero.
+///
+///     let count = COUNT.fetch_add(1, Ordering::SeqCst);
+///
+///     // Check to see if we've finished counting or not.
+///     if count < 6 {
+///         future::ready(Some(count))
+///     } else {
+///         future::ready(None)
```

Comment:
We should make this an async closure (closure of async block) instead

Reply:
Unfortunately, probably the equivalent of this cannot be written by async/await at this time. (see https://github.com/rust-lang-nursery/futures-rs/pull/1842#issuecomment-526666577)

Reply:
(This is one of the unuseful features I mentioned in https://github.com/async-rs/async-std/issues/129#issuecomment-526833077.)

Reply:
My understanding is that it can't close over futures, but it *can* access `static`s, which may make it more useful?